File: Algorithms/Trees/BinarySearchTree/js/bst.js
'use strict';

function pad(num) {
  let res = '';

  while (num--) {
    res += ' ';
  }

  return res;
}

class Node {
  constructor(value) {
    this.value = value;
    this.left = undefined
    this.right = undefined;
  }

  toJSON() {
    return {
      value: this.value,
      left: this.left,
      right: this.right
    }
  }
}

class Tree {
  add(value) {
    let node = this.root;

    if (!node) {
      this.root = new Node(value);
      return;
    }

    while (node) {
      let nodeValue = node.value;
      let dir;

      if (nodeValue > value) {
        dir = 'left';
      } else if (nodeValue < value) {
        dir = 'right';
      } else {
        break;
      }

      if (node[dir]) {
        node = node[dir];
      } else {
        node[dir] = new Node(value);
      }
    }
  }

  toJSON() {
    return this.root.toJSON();
  }
}

module.exports = Tree;

File: Algorithms/Trees/BinarySearchTree/rb/bst.rb
module BST
  class Node < Struct.new(:value, :left, :right)
    def to_json(*args)
      hash = {
        value: value
      }

      hash[:left] = self.left.to_json unless self.left.nil?
      hash[:right] = self.right.to_json unless self.right.nil?

      hash
    end
  end

  attr_reader :root

  def add(value)
    node = root

    if node.nil?
      @root = Node.new(value)
      return
    end

    while !node.nil?
      node_value = node.value
      dir = nil

      if node_value > value
        dir = :left
      elsif node_value < value
        dir = :right
      else
        return
      end

      if node[dir].nil?
        node[dir] = Node.new(value)
      else
        node = node[dir]
      end
    end
  end

  def to_json
    root.to_json
  end
end
